imp: add limited exec support

grondo opened the pull request:
This PR adds a minimal working version of the `flux-imp exec` subcommand.
This version takes input as JSON on stdin. The current format requires a single key "J", expected to contain the users signed jobspec.

`flux-imp exec` reads the path to the job shell to execute plus 1 argument from the cmdline. This allows simply inserting `flux-imp exec` in front of the job shell invocation already being done by the job-exec module.

With the exec command comes a new `[exec]` table in the IMP's configuration. In order for `flux-imp exec` to be functional, the table must contain at least the following keys
```toml
[exec]
allowed-users = [ "list of users allowed to run imp exec" ]
allowed-shells = [ "list of shell paths allowed for imp exec" ]
```

Without either of these config keys either the IMP user or the job shell will not be allowed and thus `flux imp exec` will not be functional.

Additional changes of note:
 - The flux-imp build was split into an installed IMP (`inst/flux-imp`) and test IMP `./flux-imp`. The only difference is how the IMP reads its config. The test IMP's config can be overridden at runtime, and it reads IMP and libflux-security config from the same config pattern.

Closes #96

grondo commented:
Added some build/install fixes here. `flux-imp` now gets installed to `$libexecdir/flux`, and an example `imp.toml` is added and installed to `$sysconfdir/flux/imp/conf.d`

grondo commented:
Actually, I removed the commit with an example `imp.toml` which is installed. Installing default config files from `make install` is probably not wise, since by default install_DATA files are always overwritten. For now we just won't install a imp.toml config (We'll run into the same problem with `sign.toml` eventually)

garlick commented on src/libutil/cf.c:
```diff
@@ -193,6 +193,23 @@ int cf_array_size (const cf_t *cf)
     return cf ? json_array_size (cf) : 0;
```
Typo in commit message "fucntionality"

garlick commented on src/imp/passwd.c:
```diff
@@ -0,0 +1,47 @@
+void passwd_destroy (struct passwd *pwd)
+{
+    free (pwd);
+}
```
need to free pw_name, pw_passwd

garlick commented on src/imp/passwd.h:
```diff
@@ -0,0 +1,28 @@
+/*
+ *  Return a copy of the passwd entry for UID
+ *  Caller must free with passwd_destroy()
+ */
```
Perhaps mention that some fields are not copied.

grondo replied:
Yeah, sorry meant to update to copying all fields before pushing this up. Got too excited that it was working.. next push will have a fix.

grondo commented:
> I started a review then realized that you probably weren't ready, so I'll leave off until I get the thumbs up. Leaving my comments so far in case they are useful.

Oh, yeah sorry about the bad code. The comments are indeed helpful, we should be extra careful about code in the IMP anyway.

I'll try to make another pass and get this into mergeable shape asap.

grondo commented:
I pushed a bit more work here, with some added tests to ensure IMP fails as expected when conditions for operation aren't meant.

This may be ready for another look @garlick.

Still todo is to launch the job shell with a "clean" environment, instead of inheriting from the IMP (which gets the environment propagated in `flux_rexec()` by the job-exec module, cleaned up slightly by the kernel if the IMP is setuid), but I'm not sure exactly how to do this yet.

There is a well known set of FLUX_* and PATH* environment variables that are known to be required for proper job shell initialization, but I worry about propagating this knowledge directly into flux-security, and it too easily getting out-of-sync. Similar to sudo, we could have a `env-keep` and `env-delete` configuration options, but this seems a bit onerous and also prone to error.

Perhaps for now we leave the environment as-is, since the IMP can only by run by "trusted" users, and due to setuid, truly unsafe environment variables are already scrubbed.

The other missing piece is `flux-imp kill` command, which I'll be working on today.

garlick commented on t/t2000-imp-exec.t:
```diff
@@ -0,0 +1,119 @@
+test_expect_success 'flux-imp exec returns error with bad JSON input ' '
+	echo "{}" | test_must_fail $flux_imp exec shell arg
+'
+test_expect_success 'flux-imp exec returns error with bad JSON input ' '
```
duplicate of previous test?

grondo replied:
I think I had meant to echo some invalid JSON in the 2nd test, thanks for catching this!

garlick commented on t/t2000-imp-exec.t:
```diff
@@ -0,0 +1,119 @@
+	( export FLUX_IMP_CONFIG_PATTERN=sign-none.toml &&
+	  echo foo | \
+	    test_must_fail $flux_imp exec echo good >badjson.log 2>&1
+	)
```
broken && chain?   (Same for tests following this one)

grondo replied:
Yeah, good catch, thanks! These were all cut-and-paste so I'll fix em all

garlick commented on src/imp/passwd.c:
```diff
@@ -0,0 +1,58 @@
+
+void passwd_destroy (struct passwd *pwd)
+{
+    free (pwd->pw_name);
```
might be good to handle NULL pwd here as its a common flux-core idiom to do this check in the destructor not the caller

grondo replied:
Good point!

garlick commented on src/imp/privsep.c:
```diff
@@ -164,28 +164,31 @@ privsep_t * privsep_init (privsep_child_f fn, void *arg)
+        if (ps->wfd > 0)
+            close (ps->wfd);
+        if (ps->rfd > 0)
+            close (ps->rfd);
```
I realize this is not code that originated in this PR, but since 0 is a valid file descriptor, and these values are initialized from `upfds` and `ppfds` which use -1 to designate a closed fd, should this be checking for >= 0?

grondo replied:
Yes, I think you're correct. I'll add a fix to this PR.

grondo commented:
> OK, made a pass through and have a few minor comments. This is looking really good!
(I'd be inclined to approve but wip: and your comment imply that you might have more work pendiing?)

Thanks! I addressed your comments, I think, and pushed some `fixup!` commits (though the test fixes were squashed, sorry about that).

Still todo here:

 - Better handling of job shell environment
 - Add `flux-imp kill` command. (O/w jobs started under the IMP can't be killed)

If you'd like, we could merge this minimal PR and the other two could be saved for near future PRs.

garlick commented:
Thanks, I think this could go in and further work could be in a new PR.
Squash?

On Tue, Mar 10, 2020, 11:26 AM Mark Grondona <notifications@github.com>
wrote:

> OK, made a pass through and have a few minor comments. This is looking
> really good!
> (I'd be inclined to approve but wip: and your comment imply that you might
> have more work pendiing?)
>
> Thanks! I addressed your comments, I think, and pushed some fixup!
> commits (though the test fixes were squashed, sorry about that).
>
> Still todo here:
>
>    - Better handling of job shell environment
>    - Add flux-imp kill command. (O/w jobs started under the IMP can't be
>    killed)
>
> If you'd like, we could merge this minimal PR and the other two could be
> saved for near future PRs.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-security/pull/104?email_source=notifications&email_token=AABJPW3ZWG3DRNBBSIJEWO3RG2A5PA5CNFSM4LDGI2SKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEOMTA5Q#issuecomment-597241974>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABJPW4DHD7PWJ76N2YLCKLRG2A5PANCNFSM4LDGI2SA>
> .
>

grondo commented:
Ok, squashed the fixup commits.

garlick commented:
I set merge-when-passing because there's a `.mergify.yml` in the source tree, but I had to add the label. Does that work here?

garlick commented:
Mergify seems stuck
 > Branch Protection is enabled and is preventing Mergify to merge the pull request. Mergify will merge when branch protection settings validate the pull request. (detail: 2 of 3 required status checks are failing.)

Should I just press the button or did you want to have a closer look at that @grondo?

grondo commented:
Hm, let me take a quick look. Not sure what we forgot here.

grondo commented:
Branch protection required that coverage checks passed. It is going to be increasingly difficult to keep coverage up unfortunately, so I've removed that protection.

garlick commented:
Now do we need to poke it with a stick?

grondo commented:
@mergifyio refresh

grondo commented:
@mergifyio refresh

grondo commented:
I had missed another codecov branch protection. Fixed now and looks like mergify worked.

garlick commented:
Victory!
